File: day1/dart/fun.dart
/* dart - fun.dart 
we need a main
dart does not have a repl ? oh poor boy...

*/

import 'dart:io';

var str;

int height() {
  int i ;
  int floor = 0;
  for (i =0 ; i < str.length ; i ++){
    if (str[i] == '('){
      floor ++;
    }
    else if (str[i] ==')' ){
      floor --;
    }
  }  
  return floor;
}

int basement() {
  int i ;
  int ch = 0;
  int floor = 0;
  for (i =0 ; i < str.length ; i ++){
    if (str[i] == '('){
      floor ++;
      ch ++;
    }
    else if (str[i] ==')' ){
      floor --;
      ch ++;
      if (floor < 0){
        return ch;
      }
    }
  }
  //not found
  return -1;
}

void main() async {
  try {
    final file = File('../input');
    str = await file.readAsString();
    print(str);
    print("santa reached floor ${height()}");
    print("santa basement was at character ${basement()}");
  } catch (e) {
    print('Error reading file: $e');
  }
}


/*
how does dart compare to javascript ?

santa reached floor 280
santa basement was at character 1797

*/

File: day1/javascript/fun.js

/* javascript fun.js
   how do we read contents of a file into variable ?
   backticks to interpolate a string -- assume backticks optimized away
   
*/


const fs = require('fs')

let str = fs.readFileSync('../input').toString()

console.log(str);
console.log(str.length);
   
str = str.replace(/[^()]/g, '');
console.log(str.length);
const strlen = str.length;

function height (){
    let floor = 0;
    for(let i=0;i< strlen ;i++){
	if (str[i] === '('){
	    floor ++;
	}
	else if (str[i] === ')'){
	    floor --;
	}	
    }
    return floor
}

console.log(`santa gets off on floor ${height()}`);

function basement (){
    let floor = 0;
    let ch = 0;
    for(let i=0;i< strlen ;i++){
	if (str[i] === '('){
	    floor ++;
	    ch++;
	}
	else if (str[i] === ')'){
	    floor --;
	    ch++;
	    if (floor < 0){
		return ch;
	    }
	}	
    }
    return -1;
}

console.log(`santa finds the basement at character ${basement()}`);



